### DISCUSSION TITLE
Deploying to ECS Fargate - What about Worker?

### DISCUSSION THREAD

**Discussion by jonathan-franzen** [saleor/saleor#13836] [Deployment]
Hi,

I'm currently setting up my own hosting on AWS. So far I've gotten it working, but I have a few questions regarding the setup. I have not worked with Django before, and have no experience in the worker. 

I setup the Dashboard on an S3 bucket & cloudfront. The storefront I run using AWS amplify. For the Core, the setup is as follows:

Redis: AWS elasticache
Postgres: AWS Rds
API: AWS ECS Fargate

My question is regarding the worker. Is the worker required, and if so, how/ where do i deploy it? I see that its using the Core Docker image, so this has me wondering if I can just run the worker on the same container, or if it perhaps is running automatically when I start the API container?

**Comment by daltonsena** [comment]
Hello, @jonathan-franzen 

Although the "worker" is not mandatory for you to run the core/api, I'm almost certain that without it, you won't achieve the results you're looking for. Several important asynchronous tasks are executed by it.

You should at least take a look at the [task_always_eager](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-task_always_eager) parameter (formerly CELERY_ALWAYS_EAGER). However, this is not recommended in a production environment.

Regarding ECs/Fargate, I believe you have two possible scenarios:

1 - Run multiple containers in a single service
2 - Set up a separate service just for the "worker" (in case you want to better isolate your resources)

In both cases, you can easily configure these options through the `task definition`.

Here is a simplified example of what this would look like in a scenario with multiple containers for a single "task definition":

```
{
    "taskDefinitionArn": "arn:aws:ecs:us-east....",
    "containerDefinitions": [
        {
            "name": "core",
            "image": "...dkr.ecr.us-east......",
            "cpu": X,
            "memory": Y,
            "memoryReservation": XYZ,
            "portMappings": [
                {
                    "containerPort": 8000,
                    "hostPort": 8000,
                    "protocol": "tcp"
                }
            ],
            "essential": true,
            "entryPoint": [
                "./scripts/entrypoint.sh"
            ],
            "command": [
                "gunicorn",
                "--bind",
                ":8000",
                "--workers",
                "4",
                "--worker-class",
                "saleor.asgi.gunicorn_worker.UvicornWorker",
                "saleor.asgi:application"
            ],
            "environment": [],
            "dependsOn": [
                {
                    "containerName": "worker",
                    "condition": "START"
                },
                {
                    "containerName": "redis",
                    "condition": "START"
                }
            ],
            "startTimeout": 30,
            "stopTimeout": 30,
            "readonlyRootFilesystem": false,
            "ulimits": [],
            "logConfiguration": {
                "logDriver": "awslogs",
            }
        },
        {
            "name": "worker",
            "image": "...dkr.ecr.us-east...",
            "cpu": 0,
            "portMappings": [],
            "essential": false,
            "command": [
                "celery",
                "-A",
                "saleor",
                "--app=saleor.celeryconf:app",
                "worker"
            ],
            "environment": [],
            "mountPoints": [],
            "volumesFrom": [],
            "readonlyRootFilesystem": false,
            "logConfiguration": {
                "logDriver": "awslogs",
            }
        },
        {
            "name": "redis",
            "image": "public.ecr.aws/docker/library/redis:alpine3.16",
            "cpu": 0,
            "portMappings": [
                {
                    "containerPort": 6379,
                    "hostPort": 6379,
                    "protocol": "tcp"
                }
            ],
            "essential": false,
            "environment": [],
            "mountPoints": [],
            "volumesFrom": [],
            "readonlyRootFilesystem": false,
            "logConfiguration": {
                "logDriver": "awslogs",
            }
        }
    ],
    "family": "saleor",
    "taskRoleArn": "arn:aws:iam::.....",
    "executionRoleArn": "arn:aws:iam::.....",
    "networkMode": "awsvpc",
	.
	.
	.
}
```

**Reply by daltonsena** [reply]
@jonathan-franzen By the way, I'm not on the Saleor team. They sure know more about it.

**Reply by jonathan-franzen** [reply]
@daltonsena  This reply was amazing! Thank you very much, this helped me alot. I got it working just now 👍

I ended up running them on the same task definition with different container-definitions. had to ramp up the Memory a bit, but now it is working great.
